odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2024 @@
+
+    Returns
+    -------
+    inner : `callable`
+        Inner product function with given weight. Constant weightings
+        are applicable to spaces of any size, for arrays the sizes
+        of the weighting and the space must match.
+
+    See also
+    --------
+    NumpyTensorSpaceConstWeighting, FnVectorWeighting, FnMatrixWeighting
+    """
+    return _weighting(weight, exponent=2.0).inner
+
+
+def weighted_norm(weight, exponent=2.0):
+    """Weighted norm on `Fn` spaces as free function.
+
+    Parameters
+    ----------
+    weight : scalar or `array-like`
+        Weight of the norm. A scalar is interpreted as a
+        constant weight, a 1-dim. array as a weighting vector and a
+        2-dimensional array as a weighting matrix.
+    exponent : positive float
+        Exponent of the norm. If ``weight`` is a sparse matrix, only
+        1.0, 2.0 and ``inf`` are allowed.
+
+    Returns
+    -------
+    norm : `callable`
+        Norm function with given weight. Constant weightings
+        are applicable to spaces of any size, for arrays the sizes
+        of the weighting and the space must match.
+
+    See also
+    --------
+    NumpyTensorSpaceConstWeighting
+    NumpyTensorSpaceArrayWeighting
+    """
+    return _weighting(weight, exponent=exponent).norm
+
+
+def weighted_dist(weight, exponent=2.0, use_inner=False):
+    """Weighted distance on `Fn` spaces as free function.
+
+    Parameters
+    ----------
+    weight : scalar or `array-like`
+        Weight of the distance. A scalar is interpreted as a
+        constant weight, a 1-dim. array as a weighting vector and a
+        2-dimensional array as a weighting matrix.
+    exponent : positive float
+        Exponent of the norm. If ``weight`` is a sparse matrix, only
+        1.0, 2.0 and ``inf`` are allowed.
+    use_inner : bool, optional
+        Calculate ``dist`` using the formula
+
+            ``||x - y||^2 = ||x||^2 + ||y||^2 - 2 * Re <x, y>``
+
+        This avoids the creation of new arrays and is thus faster
+        for large arrays. On the downside, it will not evaluate to
+        exactly zero for equal (but not identical) ``x`` and ``y``.
+
+        Can only be used if ``exponent`` is 2.0.
+
+    Returns
+    -------
+    dist : `callable`
+        Distance function with given weight. Constant weightings
+        are applicable to spaces of any size, for arrays the sizes
+        of the weighting and the space must match.
+
+    See also
+    --------
+    NumpyTensorSpaceConstWeighting
+    NumpyTensorSpaceArrayWeighting
+    """
+    return _weighting(weight, exponent=exponent,
+                      dist_using_inner=use_inner).dist
+
+
+def _norm_default(x):
+    """Default Euclidean norm implementation."""
+    if _blas_is_applicable(x):
+        nrm2 = linalg.blas.get_blas_funcs('nrm2', dtype=x.dtype)
+        norm = partial(nrm2, n=native(x.size))
+    else:
+        norm = np.linalg.norm
+    return norm(x.data)
```

Comment:
How are we going to go about these? Just treat them like in 1D? That would be the most practical approach.